### DISCUSSION TITLE
System.LimitException: JSON string exceeds heap size limit

### DISCUSSION THREAD

**Discussion by staggerlee99** [jamessimone/apex-rollup#368] [Q&A]
Hi all - I'm getting this error "System.LimitException: JSON string exceeds heap size limit" with just under 3,000 records. I have 5 rollups, two COUNTS, two SUMS and a LAST. For SUMS, I'm using a number field ("Counter__c") on each row that defaults to "1".
Here are the where clauses with the roll-up type in ():

- **Stage__c = 'Rejected'**  (COUNT)
- **Stage__c NOT IN ('New', 'Offer Sent', 'Offer Accepted', '98S Denied', 'Rejected')** (SUM) 
- **Stage__c = '98S Denied'** (COUNT)
- **Stage__c <> '98S Denied' AND Offer_Accepted_or_Rejected__c != NULL** (LAST)
- The other "SUM" simply adds up all the records in the dataset, using the value in Counter__c. I've changed this to just a simple COUNT roll-up type, but I still get this error. 

Any insights on what I may be doing incorrectly would be greatly appreciated!
-Tom

**Comment by jamessimone** [comment]
Hey Tom, that's an interesting error and I'd be curious to hear where exactly you're running into it. If you have your Apex Developer Console open while running rollups, for example, you should be able to search for that `LimitException` within the log that has the error to see the full stacktrace. Additionally - is this coming up when you're running a full recalc from the app, or from some other operation?

In particular, I'm curious if this is happening with Rollup Logging Enabled (from your `Rollup Control` CMDT record) or by some other means. Either way, it should be a relatively easy fix once we've identified where it's happening.

**Reply by staggerlee99** [reply]
Hi James! Thanks for your reply. I'm running it during a scheduled flow. I do have Rollup Logging Enabled in the Rollup Control metadata.  Here are excerpts from a log - hope this helps!

- (System Code)
Class.RollupFlowBulkProcessor.addRollup: line 134, column 1
13:35:48.340 (6341029506)|FATAL_ERROR|System.LimitException: JSON string exceeds heap size limit

- 3:35:48.340 (6341533346)|CODE_UNIT_FINISHED|RollupFlowBulkProcessor.addRollup(List<RollupFlowBulkProcessor.FlowInput>)
13:35:48.963 (6963870086)|FLOW_ELEMENT_ERROR|An Apex error occurred: System.LimitException: JSON string exceeds heap size limit

**Reply by jamessimone** [reply]
@staggerlee99 I would try disabling rollup logging. With that many records - and a few rollups to go through - it's possible that the heap size is being exceeded due to logging (especially if the logging level you're using is more than `DEBUG`). In fact, I see the likely culprit as I'm looking at the code right now. I can optimize that particular piece of logging to be more bulk record-friendly and release the fix for that in the next version, but I think you should be good in the meantime with logging disabled.

**Reply by staggerlee99** [reply]
Fantastic. Trying that now...

**Reply by staggerlee99** [reply]
Sorry - failed with the same error. I turned off logging and ran the scheduled flow in debug mode. Wish I had better news :-).

**Reply by jamessimone** [reply]
@staggerlee99 no worries. Are you using the CMDT-driven rollup flow action, or the base flow action? There's another potentially problematic line in the CMDT-driven one that I may also need to update (which I can provide you with guidance on updating in a sandbox, if you're interested).

**Comment by staggerlee99** [comment]
Thanks, James. I'm using the "Perform Rollup__mdt-based rollup" action. Does that answer your question? Also - this is followed by the "Process Deferred Rollups" action in the Flow. The Flow errors-out before it gets to the Process Deferred Rollups action.

**Reply by jamessimone** [reply]
Yup, that answers my question. Can you try updating the following?

You would update this line in `RollupFlowBulkProcessor`:

(current line):
```java
output.message = 'Rollup queued for context: ' + meta.RollupOperation__c + ' with metadata values: ' + JSON.serialize(input);
```

(updated line):
```java
output.message = 'Rollup queued for context: ' + meta.RollupOperation__c;
```

**Reply by staggerlee99** [reply]
Yes - will do - won't be for a few hours. I will send you an update ASAP. Thank you!!

**Reply by staggerlee99** [reply]
Hi - sorry if this is an obvious error/fix: I updated the code in a Sandbox, but I cannot push it to prod since the class is part of managed package. Is there a workaround for that? I'll be away for the next hour or so, FYI...thanks again for your help with this!

**Reply by jamessimone** [reply]
@staggerlee99 I will be including all of these optimizations in the next version; unfortunately you would not be able to deploy the changes we have been talking about without that. I will try to wrap up the work I am doing and get that updated version out there today.

**Reply by staggerlee99** [reply]
That would be great. Thanks soooo much!

**Reply by jamessimone** [reply]
@staggerlee99 this should be all set with the updated version now available!

**Reply by staggerlee99** [reply]
Amazing! Will install in the morning. THANK YOU!!!

**Comment by staggerlee99** [comment]
Hi James - I installed the update this morning -- and, while the Heap Size issue went away I'm back to the initial error  - an Apex timeout. I've attached the log. If you get a chance, could you take a look and let me know what you think may be causing the error? Can't say thanks enough!
[apex-07L7V00005KhI5PUAV.log](https://github.com/jamessimone/apex-rollup/files/9716883/apex-07L7V00005KhI5PUAV.log)

**Comment by jamessimone** [comment]
What do you have set on the Org Default record for the Batch Chunk Size and Max Lookup Rows Before Batching?

I would try lowering the numbers for both those fields. There can be CPU timeouts when updating parent records if there’s more automation in the org in question for updates made there; typically tweaking those amounts is enough to get everything working smoothly.

I won’t be able to take a detailed look at the log for a few days, but let me know if the above suggestions work and otherwise I’ll be happy to take a look as soon as I am able. Thanks!

**Reply by staggerlee99** [reply]
Thanks, James. I've been tweaking those values, right now they are set to the following:
Max Lookup Rows Before Batching: 250
Batch Chunk Size:100
Max Number Of Queries: 50
Max Parent Rows Updated At Once: 250
Max Rollup Retries: 20

**Reply by staggerlee99** [reply]
Thanks again for taking the time to help us out with this!

**Reply by jamessimone** [reply]
@staggerlee99 that log you attached seems to only be the synchronous part of the log - which seems to end successfully. Do you still have `Is Rollup Logging Enabled` set to true? What logging level do you have set on the `Rollup Plugin Parameter` for Logging Debug Level? What logging level do you have set in the Salesforce Developer Console's trace for your user? All of these things will impact CPU time - and you should see the log where the timeout is occurring (either a QueueableHandler log, or a Batch log)
